### PR TITLE
Connect to Opera Wallet button added

### PR DESCRIPTION
**Pull request by filipkozakred**
## Description

Closes # (NO_ISSUE)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [x] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

**Review by guanbinrui — DISMISSED** [review]
Awesome work! Did you try to send any transaction with this new wallet?

**Comment by filipkozakred** [comment]
> Awesome work! Did you try to send any transaction with this new wallet?

Yes, it successfully connects with built in Opera Crypto Wallet

File: packages/shared-base/src/Site/index.ts [hunk 1]
```diff
@@ -45,3 +45,7 @@ export function isEnhanceableSiteType() {
 export function isExtensionSiteType() {
     return !!getExtensionSiteType()
 }
+
+export function getBrowserType() {
```
Comment (guanbinrui) [inline packages/shared-base/src/Site/index.ts]: I think we don't need to add this. Just match it inline.

**Review by guanbinrui — DISMISSED** [review]

File: packages/web3-shared/evm/constants/descriptors.ts [hunk 2]
```diff
@@ -312,6 +312,21 @@ export const PROVIDER_DESCRIPTORS: Array<ProviderDescriptor<ChainId, ProviderTyp
...
+    {
+        ID: `${PLUGIN_ID}_opera`,
+        providerAdaptorPluginID: PLUGIN_ID,
+        type: ProviderType.Opera,
+        name: 'Opera',
+        icon: new URL('../assets/opera.png', import.meta.url),
+        enableRequirements: {
+            supportedChainIds: getEnumAsArray(ChainId).map((x) => x.value),
+            supportedEnhanceableSites: getEnumAsArray(EnhanceableSite).map((x) => x.value),
+            supportedExtensionSites: getEnumAsArray(ExtensionSite).map((x) => x.value),
```
Comment (guanbinrui) [inline packages/web3-shared/evm/constants/descriptors.ts]: Oops, I found an issue here after approving this PR. Not all EVM-compatible chains are supported. And please disable it for extension-added sites.

image

**Comment by AinuX** [comment]
@filipkozakred 

1. Opera crypto wallet also support Polygon, please add it as an option when select Polygon chain
image

2. Fail to connect BNB chain with opera wallet.

https://user-images.githubusercontent.com/2582974/179174539-a9690e6d-d064-48a5-80dd-06839b721107.mp4


3. Fail to switch chain even switch to supported chain.
4. Since Opear wallet only support ETH, BNB Chain and Polygon, it's better to disable switch to chain unsupported chain and give a reminder message to enduser.

https://user-images.githubusercontent.com/2582974/179174058-793c73c9-6b21-4a0b-9a45-eae528cc6c17.mp4


like this,
image



5. Lost connection after sign message/tx

https://user-images.githubusercontent.com/2582974/179175955-05ce29ef-2e1e-4cb4-a21f-3c4833f28f53.mp4



6.  Unable to unlock a token 

https://user-images.githubusercontent.com/2582974/179178695-e95169d6-a6cb-448a-b3d3-4ed56abb3a46.mp4

File: packages/plugins/EVM/src/state/Connection/providers/Opera.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    override get readyPromise() {
+        const isOpera = window.navigator.userAgent.includes('OPR/')
+        if (!isOpera) {
+            return Promise.reject(new Error('Not available on this browser.'))
+        }
+        return this.bridge.untilAvailable().then(() => undefined)
+    }
+    override async switchChain(chainId?: ChainId) {
+        await SharedContextSettings.value.updateAccount({
```
Comment (guanbinrui) [inline packages/plugins/EVM/src/state/Connection/providers/Opera.ts]: The shared context serves only for MaskWallet. For Opera, I think we can do:

1. Check the given `chainId` is supported by Opera. If not, throw a deny error.
2. If the expected `chainId` is not the same as the current one. Reconnect it `this.connect(chainId)`.

**Comment by guanbinrui** [comment]
Hi @filipkozakred FYI, we have integrated some UI components for testing here.

image
image

File: packages/plugins/EVM/src/state/Connection/providers/Opera.ts (hunk 3 above)
Comment (filipkozakred) [reply packages/plugins/EVM/src/state/Connection/providers/Opera.ts]: I tried to do it this way. But unfortunately it didn't work. Changed the chain in Maskbook, but did not change the chain in Opera Crypto Wallet. This change I made works surprisingly flawlessly. Does it need to be changed?

**Pull request closed without merging**